### PR TITLE
Support package upgrade in for the RDMA driver

### PR DESCRIPTION
**Pull request by rjschwei**
+ A different version of the rdma driver may need to be installed when
    a VM gets rebooted as it may be deployed on different HW that has a
    different firmware version

### CODE REVIEW

**Comment by ahmetalpbalkan** [comment]
manually fixing the CLA label

File: azurelinuxagent/distro/suse/rdma.py [hunk 1]
```diff
@@ -48,22 +51,35 @@ def install_driver(self):
...
+                    logger.info(info_msg % (package_name, version))
+                    return True
+                if installed == 'i':
+                    driver_package_installed = True
+
+        # If we get here the driver package is installed but the
+        # version doesn't match or no package is installed
+        if driver_package_installed:
+            self.remove_driver_module()
+            cmd = '%s &s' % (zypper_remove, package_name)
```
Comment (ahmetalpbalkan) [inline azurelinuxagent/distro/suse/rdma.py]: %s?

File: azurelinuxagent/distro/suse/rdma.py [hunk 2]
```diff
@@ -35,10 +35,13 @@ def install_driver(self):
             logger.error(error_msg)
             return
         zypper_install = 'zypper -n in'
+        zypper_remove = 'zypper -n rm'
```
Comment (ahmetalpbalkan) [inline azurelinuxagent/distro/suse/rdma.py]: it would be good to add `%s` to commands up here.

**Comment by ahmetalpbalkan** [comment]
Change looks good to me generally. Left a few comments.

@rjschwei does suse use  `waagent.conf` config keys:
	• OS.CheckRdmaDriver 
	• OS.UpdateRdmaDriver
	• OS.RdmaRepository
	• OS.PackageExtraConfig

File: azurelinuxagent/distro/default/rdma.py [hunk 3]
```diff
@@ -21,10 +21,14 @@
...
 import os
 import re
-import azurelinuxagent.logger as logger
+import azurelinuxagent.common.logger as logger
+import azurelinuxagent.common.utils.shellutil as shellutil
 
 
 class RDMAHandler(object):
+    def __init__(self):
+        self.driver_module_name = 'hv_network_direct'
```
Comment (ahmetalpbalkan) [inline azurelinuxagent/distro/default/rdma.py]: can we declare this just under `class RDMAHander`

File: azurelinuxagent/distro/suse/rdma.py [hunk 4]
```diff
@@ -35,10 +35,13 @@ def install_driver(self):
...
         zypper_search = 'zypper se -s'
```
Comment (ahmetalpbalkan) [inline azurelinuxagent/distro/suse/rdma.py]: maybe extract these `zypper` commands just below `class RDMAHandler` (and have trailing `%s` so we can format easier)?

File: azurelinuxagent/distro/default/rdma.py [hunk 5]
```diff
@@ -73,3 +78,9 @@ def install_driver(self):
            be overwritten in the child implementation."""
 
         raise Exception('RDMAHandler.install_driver not implemented')
+
+    def remove_driver_module(self):
+        """Force remove the kernel driver"""
+        # Force remove the loaded driver
```
Comment (ahmetalpbalkan) [inline azurelinuxagent/distro/default/rdma.py]: cosmetic: this line is not very necessary

**Comment by ahmetalpbalkan** [comment]
also one general comment: `install_driver()` is getting long, if we could split that into multiple functions and clarify the flow in the  `install_driver()` better, that would be more favorable.

thanks @rjschwei.

**Comment by rjschwei** [comment]
On 06/16/2016 07:08 PM, Ahmet Alp Balkan wrote:
> Change looks good to me generally. Left a few comments.
> 
> @rjschwei does suse use  `waagent.conf` config keys:
> 	• OS.CheckRdmaDriver 
> 	• OS.UpdateRdmaDriver
> 	• OS.RdmaRepository
> 	• OS.PackageExtraConfig 
> 

No, these are unnecessary. I was planning on introducing a new variable

OS.EnableRDMA

We can skip the variable and simply look at /var/lib/hyperv/.kvp_pool_0
and do everything automatically, but I am leaning towards having the
config variable to enable the mechanism.


-- 
Robert Schweikert                   MAY THE SOURCE BE WITH YOU
Public Cloud Architect                         LINUX
rjschwei@suse.com
IRC: robjo

File: azurelinuxagent/distro/default/rdma.py (hunk 3 above)
Comment (rjschwei) [inline azurelinuxagent/distro/default/rdma.py]: On 06/16/2016 07:11 PM, Ahmet Alp Balkan wrote:
>>  
>>  
>>  class RDMAHandler(object):
>> +    def __init__(self):
>> +        self.driver_module_name = 'hv_network_direct'
> 
> can we declare this just under `class RDMAHander`

Yes, it's a style choice. I generally prefer using __init__ but if you'd
rather me drop __init__ I certainly have no problem doint it the other
way. From a language perspective I am not certain if there is actually a
difference.

Comment (rjschwei) [inline azurelinuxagent/distro/default/rdma.py]: OK after refreshing my memory a bit I agree that a class member is a nicer choice for the driver_module name

File: azurelinuxagent/distro/suse/rdma.py (hunk 4 above)
Comment (rjschwei) [inline azurelinuxagent/distro/suse/rdma.py]: Trailing %s in command set up makes sense. Having played around a bit with refactoring the whole method and ending up in a tangled mess I see no advantage to have the commands as class or instance members, thus I think we should leave them as they are with the added "%s"

**Comment by rjschwei** [comment]
Agreed that install_driver() is getting a bit lengthy. I did fiddle around with it to see if I could find clean lines of refactoring the code but I ended up in a tangled mess that I did not like at all. That is not to say that refactoring along nice lines is not possible just that I have not found a nice solution to the problem yet. THus I'd prefer to keep the method in place as is.

**Comment by rjschwei** [comment]
Testing changes....

**Comment by rjschwei** [comment]
OK, changes ready for review.

I needed to make a change to the "package upgrade" path. It turns out that this particular module will not remove with the "rmmod" command. Thus is the module is loaded into kernel space but an upgrade is needed we have no choice but to reboot the system.

Implemented the reboot condition. The question is, is it safe to do this at this point, or should this be handled at the end of the agent workflow?

**Comment by ahmetalpbalkan** [comment]
> Implemented the reboot condition. The question is, is it safe to do this at this point, or should this be handled at the end of the agent workflow?

Hi @rjschwei, any ideas why the old code in 2.0 branch have did not have any `reboot`s?

File: azurelinuxagent/distro/suse/rdma.py [hunk 6]
```diff
@@ -34,11 +34,14 @@ def install_driver(self):
             error_msg += 'no driver will be installed.'
             logger.error(error_msg)
             return
-        zypper_install = 'zypper -n in'
-        zypper_search = 'zypper se -s'
+        zypper_install = 'zypper -n in %s'
+        zypper_remove = 'zypper -n rm %s'
+        zypper_search = 'zypper se -s %s'
```
Comment (ahmetalpbalkan) [inline azurelinuxagent/distro/suse/rdma.py]: In the old code we had:

     self.command_to_get_rdma_package_info = self.rpm_path + " -q " + self.rdma_package_name

which would just return the package name like `microsoft-hyper-v-rdma-4.1.0.142-20160323.x86_64`. Isn't this a better fit to determine the installed package?

**Comment by ahmetalpbalkan** [comment]
> No, these are unnecessary. I was planning on introducing a new variable
> 
> OS.EnableRDMA
> 
> We can skip the variable and simply look at /var/lib/hyperv/.kvp_pool_0
> and do everything automatically, but I am leaning towards having the
> config variable to enable the mechanism.

@rjschwei regarding these variables I mentioned earlier, do we know that customers were not using them in their custom images at all? Because it had pre-post package install script hooks, extra repostories and packages to install. 

If we don't provide the same keys, it might mean regression for the new users.

Here is the logic I extracted from 2.0:

```
read config keys from waagent.conf
	• OS.CheckRdmaDriver 
	• OS.UpdateRdmaDriver
	• OS.RdmaRepository
	• if OS.PackageExtraConfig: read from JSON:
			§ PreTask (file path)
			§ PostTask (file path)
			§ ExtraRepositories (array)
			§ Packages (array)

if $OS.RdmaRepository:
	distro.addPackageRepository(...)

exec $PreTask  [continue on failure]

if $OS.UpdateRdmaDriver:
	distro.rdmaUpdate() [continue on failure]
	
foreach $ExtraRepositories:
	distro.addPackageRepository(...)

foreach $Packages:
	distro.refreshRepository()
	distro.installPackage(...)  [continue on failure]

exec $PostTask  [continue on failure]

if $OS.CheckRdmaDriver:
	res = distro.checkRDMA()
        print res
```

**Comment by rjschwei** [comment]
No idea why the 2.0 code did not have the reboot, in the case where a VM would have been relaunched  on HW with a different FW implementation the 2.0 branch would have left the user in a state where IB would not have been usable as the wrong driver would have been loaded in the kernel.

File: azurelinuxagent/distro/suse/rdma.py (hunk 6 above)
Comment (rjschwei) [inline azurelinuxagent/distro/suse/rdma.py]: Well it is not necessarily about finding what is installed, it is also about finding what is available. The "zypper se -s" call gives us both. We have to cover 3 conditions:

1.) No driver installed => install the proper driver version from the available packages, thus I have to find all the avialbale packages
2.) Driver installed and matching version => do nothing (this case could be covered by the rpm call from the 2.0 branch
3.) Driver installed but the FW version does not match => uninstall the driver and find the proper version from all the available packages

All cases are covered by the information zypper returns. Only 1 case is covered by the information the rpm call returns.

**Comment by rjschwei** [comment]
So the repository is already built into the image and if it doe not work because the system sits in a VNET without Internet access then we have the local cache of packages that is alos built into the image. This makes OS.RdmaRepository obsolete. 

With the new logic the driver automatically gets updated when needed, this makes OS.UpdateRdmaDriver obsolete.

The use of OS.CheckRdmaDriver is replaced by the new setting OS.EnableRDMA with the new setting being more explicit in name as it indicates that the RDMA driver will actually be enabled. The previous name only indicates that the driver would be checked.

OS.PackageExtraConfig is IMHO not related to RDMA, if it is then from my perspective it is intermingling of different concepts. I can see that there may be value in having OS.PackageExtraConfig, but I would argue that any functionality hidden behind this should be executed no matter what and that the switch is superfluous.

**Comment by ahmetalpbalkan** [comment]
@rjschwei sounds good. I realized CentOS did not really use these flags either. I am assuming they're added there for users to extend. Let's leave them out for now and see what happens.

File: azurelinuxagent/distro/suse/rdma.py [hunk 7]
```diff
@@ -48,22 +51,42 @@ def install_driver(self):
...
+                result = shellutil.run(cmd)
+                if result:
+                    error_msg = 'Failed install of package "%s" '
+                    error_msg += 'from available repositories.'
+                    logger.error(error_msg % complete_name)
+                msg = 'Successfully installed "%s" from '
+                msg += 'configured repositories'
+                logger.info(msg % complete_name)
+                self.load_driver_module()
+                if requires_reboot:
```
Comment (ahmetalpbalkan) [inline azurelinuxagent/distro/suse/rdma.py]: can we please add some logging around the reboot. it is quite a critical operation for us?

File: azurelinuxagent/distro/suse/rdma.py (hunk 6 above)
Comment (ahmetalpbalkan) [inline azurelinuxagent/distro/suse/rdma.py]: @rjschwei sounds good.

**Comment by ahmetalpbalkan** [comment]
I guess most of the `install_driver()` logic is going to be the same for CentOS. including checking for FW version, matching it against installed (if any), upgrading it etc. The installing from `/opt/microsoft/rdma` is going to be the same as well. Only the actual commands and the way we parse versions etc will differ.

I guess we can generalize that method more later on.

**Comment by ahmetalpbalkan** [comment]
LGTM

**Pull request merged**